src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -53,10 +54,15 @@ internal sealed class PrivateImplementationDetails : DefaultTypeDef, Cci.INamesp
         // Once frozen the collections of fields, methods and types are immutable.
         private int _frozen;
 
-        // fields mapped to metadata blocks
         private ImmutableArray<SynthesizedStaticField> _orderedSynthesizedFields;
-        private readonly ConcurrentDictionary<ImmutableArray<byte>, MappedField> _mappedFields =
-            new ConcurrentDictionary<ImmutableArray<byte>, MappedField>(ByteSequenceComparer.Instance);
+
+        // fields mapped to metadata blocks
+        private readonly ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment), MappedField> _mappedFields =
+            new ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment), MappedField>(MappedFieldEqualityComparer.Instance);
+
+        // fields for cached arrays
+        private readonly ConcurrentDictionary<(ImmutableArray<byte> Data, SpecialType ElementType), CachedArrayField> _cachedArrayFields =
+            new ConcurrentDictionary<(ImmutableArray<byte> Data, SpecialType ElementType), CachedArrayField>(CachedArrayFieldEqualityComparer.Instance);
```

Comment:
>SpecialType [](http://example.com/codeflow?start=65&length=11)

I am fine with using ```SpecialType``` here, but there is an option to use ```ushort``` here and share the comparer, if you are after reducing number of types.